tsl/src/nodes/decompress_chunk/decompress_chunk.c

```diff
@@ -360,6 +370,196 @@ cost_decompress_chunk(Path *path, Path *compressed_path)
+ *       column: [0, 3] [0, 5] [3, 7] [6, 10]
+ *
+ *   (2) The decompress chunk node initializes a binary heap, opens the first batch and
+ *       decompresses the first tuple from the batch. The tuple is put on the heap. In addition
+ *       the opened batch is marked as the most recent batch (MRB).
+ *
+ *   (3) As soon as a tuple is requested from the heap, the following steps are performed:
+ *       (3a) If the heap is empty, we are done.
+ *       (3b) The top tuple from the heap is taken. It is checked if this tuple is from the
+ *            MRB. If this is the case, the next batch is opened, the first tuple is decompressed,
+ *            placed on the heap and this batch is marked as MRB. This is repeated until the
+ *            top tuple from the heap is not from the MRB. After the top tuple is not from the
+ *            MRB, all batches (and one ahead) which might contain the most recent tuple are
+ *            opened and placed on the heap.
+ *
+ *            In the example above, the first three batches are opened because the first two
+ *            batches might contain tuples with a value of 0.
+ *       (3c) The top element from the heap is removed, the next tuple from the batch is
+ *            decompressed (if present) and placed on the heap.
+ *       (3d) The former top tuple of the heap is returned.
+ *
+ * This function checks if the compression 'order by' and the query 'order by' are
+ * compatible and the optimization can be used.
+ */
+static MergeBatchResult
+can_batch_merge_append(PlannerInfo *root, CompressionInfo *info, Chunk *chunk)
+{
+	PathKey *pk;
+	Var *var;
+	Expr *expr;
+	char *column_name;
+	List *pathkeys = root->query_pathkeys;
+	FormData_hypertable_compression *ci;
+	MergeBatchResult merge_result = SCAN_FORWARD;
+
+	/* Ensure that we have path keys and the chunk is ordered */
+	if (pathkeys == NIL || ts_chunk_is_unordered(chunk) || ts_chunk_is_partial(chunk))
+		return MERGE_NOT_POSSIBLE;
+
+	int nkeys = list_length(pathkeys);
+
+	/*
+	 * Loop over the pathkeys of the query. These pathkeys need to match the
+	 * configured compress_orderby pathkeys.
+	 */
+	for (int pk_index = 0; pk_index < nkeys; pk_index++)
+	{
+		pk = list_nth(pathkeys, pk_index);
+		expr = find_em_expr_for_rel(pk->pk_eclass, info->chunk_rel);
+
+		if (expr == NULL || !IsA(expr, Var))
+			return MERGE_NOT_POSSIBLE;
+
+		var = castNode(Var, expr);
+
+		if (var->varattno <= 0)
+			return MERGE_NOT_POSSIBLE;
+
+		column_name = get_attname(info->chunk_rte->relid, var->varattno, false);
+		ci = get_column_compressioninfo(info->hypertable_compression_info, column_name);
+
+		if (ci->orderby_column_index != pk_index + 1)
+			return MERGE_NOT_POSSIBLE;
+
+		/* Check order, if the order of the first column do not match, switch to backward scan */
+		Assert(pk->pk_strategy == BTLessStrategyNumber ||
+			   pk->pk_strategy == BTGreaterStrategyNumber);
+
+		if (pk->pk_strategy != BTLessStrategyNumber)
```

Comment:
Might be more concise to determine the direction required by pathkey, and compare it to the previous direction we had. Like we did for compression: https://github.com/timescale/timescaledb/blob/main/tsl/src/compression/compression.c#L269-L284
Now you're basically writing out the branches for each of four possible directions.

Reply:
The code structure is similar to https://github.com/timescale/timescaledb/blob/4a6650d170cc07dbec0afe01bb3685a2cf9d9a37/tsl/src/nodes/decompress_chunk/decompress_chunk.c#L1575 . I would prefer to be more verbose here (as in the current implementation) to make the cases easier to understand.

Reply:
OK, maybe it's a matter of taste, for me this looks more complicated :)

Reply:
Personally I think the more verbose version is easier to grasp.